src/Compilers/Core/Portable/MetadataReader/PEModule.cs

```diff
@@ -1312,37 +1314,195 @@ private ArrayBuilder<string> ExtractStringValuesFromAttributes(List<AttributeInf
             return result;
         }
 
-        private ObsoleteAttributeData TryExtractObsoleteDataFromAttribute(AttributeInfo attributeInfo)
+#nullable enable
+        private ObsoleteAttributeData? TryExtractObsoleteDataFromAttribute(AttributeInfo attributeInfo)
         {
             Debug.Assert(attributeInfo.HasValue);
-
+            ObsoleteAttributeData? obsoleteData;
             switch (attributeInfo.SignatureIndex)
             {
                 case 0:
                     // ObsoleteAttribute()
-                    return new ObsoleteAttributeData(ObsoleteAttributeKind.Obsolete, message: null, isError: false);
-
+                    return TryExtractParameterlessObsoleteData(attributeInfo.Handle);
                 case 1:
                     // ObsoleteAttribute(string)
-                    string message;
-                    if (TryExtractStringValueFromAttribute(attributeInfo.Handle, out message))
+                    TryExtractValueFromAttribute(attributeInfo.Handle, out obsoleteData, s_attributeObsoleteDataExtractorString);
+                    return obsoleteData;
+                case 2:
+                    // ObsoleteAttribute(string, bool)
+                    TryExtractValueFromAttribute(attributeInfo.Handle, out obsoleteData, s_attributeObsoleteDataExtractorStringBool);
+                    return obsoleteData;
+                default:
+                    throw ExceptionUtilities.UnexpectedValue(attributeInfo.SignatureIndex);
+            }
+        }
+
+        private ObsoleteAttributeData? TryExtractParameterlessObsoleteData(CustomAttributeHandle handle)
+        {
+            Debug.Assert(!handle.IsNil);
+            try
+            {
+                BlobHandle valueBlob = GetCustomAttributeValueOrThrow(handle);
+                if (!valueBlob.IsNil)
+                {
+                    BlobReader reader = MetadataReader.GetBlobReader(valueBlob);
+
+                    if (reader.Length >= 4)
```

Comment:
>if (reader.Length >= 4) [](start = 20, length = 23)

If the length is exactly 4, do we need to try cracking the blob? Can we simply produce the result right away?

Reply:
I think in the case of exactly 4 bytes we could just verify the prologue and not call CrackObsoleteProperties.